File: c/problem_59.c
#include "problem_59.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

#define FS_CAP 16
typedef struct { char names[FS_CAP][64]; char contents[FS_CAP][256]; int size; } FileSyncer59;

static void fs_put(FileSyncer59 *f, const char *name, const char *content) {
    for (int i=0;i<f->size;i++) if(strcmp(f->names[i],name)==0){ strncpy(f->contents[i],content,255); return; }
    strncpy(f->names[f->size], name, 63);
    strncpy(f->contents[f->size], content, 255);
    f->size++;
}

static unsigned long fs_hash(const char *s) {
    unsigned long h = 5381;
    while (*s) h = h*31 + (unsigned char)*s++;
    return h;
}

static unsigned long fs_checksum(FileSyncer59 *f, const char *name) {
    for (int i=0;i<f->size;i++) if(strcmp(f->names[i],name)==0) return fs_hash(f->contents[i]);
    return 0;
}

static int fs_sync(FileSyncer59 *dst, FileSyncer59 *src, char changed[][64]) {
    int n = 0;
    for (int i=0;i<src->size;i++) {
        if (fs_checksum(dst, src->names[i]) != fs_checksum(src, src->names[i])) {
            fs_put(dst, src->names[i], src->contents[i]);
            if (changed) strcpy(changed[n], src->names[i]);
            n++;
        }
    }
    return n;
}

void problem59(void) {
    printf("Problem 59:\n");
    FileSyncer59 a={0}, b={0};
    fs_put(&a,"foo.txt","hello world"); fs_put(&a,"bar.txt","same content");
    fs_put(&b,"bar.txt","same content");
    char changed[8][64];
    int n = fs_sync(&b, &a, changed);
    assert(n==1 && strcmp(changed[0],"foo.txt")==0);
    int found=0;
    for(int i=0;i<b.size;i++) if(strcmp(b.names[i],"foo.txt")==0 && strcmp(b.contents[i],"hello world")==0) found=1;
    assert(found);
    printf("All Tests Passed\n");
}

File: rust/problem_59.rs
use std::collections::HashMap;

fn simple_hash(s: &str) -> u64 {
    s.bytes().fold(0u64, |h, b| h.wrapping_mul(31).wrapping_add(b as u64))
}

pub struct FileSyncer59 { pub files: HashMap<String, String> }

impl FileSyncer59 {
    pub fn new() -> Self { FileSyncer59 { files: HashMap::new() } }
    pub fn put(&mut self, name: &str, content: &str) { self.files.insert(name.to_string(), content.to_string()); }
    fn checksum(&self, name: &str) -> u64 { simple_hash(self.files.get(name).map(|s| s.as_str()).unwrap_or("")) }
    pub fn sync_from(&mut self, other: &FileSyncer59) -> Vec<String> {
        let mut transferred = vec![];
        for (name, content) in &other.files {
            if self.checksum(name) != other.checksum(name) {
                self.files.insert(name.clone(), content.clone());
                transferred.push(name.clone());
            }
        }
        transferred
    }
}

pub fn problem59() {
    println!("Problem 59:");
    let mut a = FileSyncer59::new(); let mut b = FileSyncer59::new();
    a.put("foo.txt", "hello world"); a.put("bar.txt", "same content");
    b.put("bar.txt", "same content");
    let mut changed = b.sync_from(&a);
    changed.sort();
    assert_eq!(changed, vec!["foo.txt"]);
    assert_eq!(b.files.get("foo.txt").unwrap(), "hello world");
    println!("All Tests Passed");
}
